File: Aminov/1261.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class FindElements:
    def recover(self, cursor: TreeNode, new_val: int):
        self.vals.append(new_val)
        if cursor.left:
            self.recover(cursor.left, 2 * new_val + 1)
        if cursor.right:
            self.recover(cursor.right, 2 * new_val + 2)

    def __init__(self, root: Optional[TreeNode]):
        self.vals = []
        self.recover(root, 0)

    def find(self, target: int) -> bool:
        return target in self.vals

# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)

File: Ивашечкин/1261.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type FindElements struct {
  	data map[int]bool
}



func Constructor(root *TreeNode) FindElements {
  	data := make(map[int]bool)
		root.Val = 0
 		data = recovery(root, data)
  	return FindElements{data: data}
}

func recovery(root *TreeNode, data map[int]bool) map[int]bool {
  	if root == nil {
				return data
		}
		data[root.Val] = true
		if root.Left != nil {
				root.Left.Val = root.Val*2 + 1
				data = recovery(root.Left, data)
		}
		if root.Right != nil {
				root.Right.Val = root.Val*2 + 2
				data = recovery(root.Right, data)
		}
		return data
}


func (this *FindElements) Find(target int) bool {
    return this.data[target]
}


/**
 * Your FindElements object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Find(target);
 */